gtest/src/wasm_executor.rs

```diff
@@ -0,0 +1,352 @@
+use gear_core::{
+    env::{Ext as ExtTrait, ExtCarrier},
+    ids::ProgramId,
+    memory::{AllocationsContext, PageBuf, PageNumber, WasmPageNumber},
+    message::{IncomingMessage, MessageContext},
+    program::Program,
+};
+use std::{collections::BTreeMap, mem};
+use wasmtime::{
+    Config, Engine, Extern, Func, Instance, Memory as WasmtimeMemory, MemoryType, Module, Store,
+    Val,
+};
+
+/// Binary meta-functions executor for testing purposes
+pub(crate) struct WasmExecutor {
+    instance: Instance,
+    store: Store<StoreData<Ext>>,
+    memory: WasmtimeMemory,
+}
+
+impl WasmExecutor {
+    /// Creates a WasmExecutor instance from a program.
+    /// Also uses provided memory pages for future execution
+    #[allow(dead_code)]
+    pub(crate) fn new(
+        source: ProgramId,
+        program: &Program,
+        memory_pages: &BTreeMap<PageNumber, Box<PageBuf>>,
+        message: Option<IncomingMessage>,
+    ) -> Self {
+        let ext = WasmExecutor::build_ext(source, program, message);
+        let ext_carrier = ExtCarrier::new(ext);
+        let store_data = StoreData {
+            ext: ext_carrier.cloned(),
+            termination_reason: None,
+        };
+
+        let config = Config::new();
+        let engine = Engine::new(&config).expect("Failed to create engine");
+        let mut store = Store::<StoreData<Ext>>::new(&engine, store_data);
+
+        let module = Module::new(&engine, program.code().code()).expect("Failed to create module");
+
+        let mut memory =
+            WasmtimeMemory::new(&mut store, MemoryType::new(program.static_pages().0, None))
+                .expect("Failed to create memory");
+        let funcs = funcs_tree::build(&mut store, memory, None);
+        let mut externs = Vec::with_capacity(module.imports().len());
+        for import in module.imports() {
+            if import.module() != "env" {
+                panic!("Non environment import in module");
+            }
+            match import.name() {
+                Some("memory") => externs.push(Extern::Memory(memory)),
+                Some(key) => {
+                    if funcs.contains_key(key) {
+                        externs.push(funcs[key].into())
+                    } else {
+                        panic!("Wasm is asking for unknown function: {:?}. Consider to add in from FuncsHandler", key)
+                    }
+                }
+                _ => continue,
+            };
+        }
+
+        let instance =
+            Instance::new(&mut store, &module, &externs).expect("Failed to create instance");
+        WasmExecutor::set_pages(&mut store, &mut memory, memory_pages)
+            .expect("Failed to set memory pages");
+
+        Self {
+            instance,
+            store,
+            memory,
+        }
+    }
+
+    /// Executes non-void function by provided name.
+    /// Panics if no function with such name was found or function was void
+    #[allow(dead_code)]
```

Comment:
remove all dead codes from everywhere, please
if any related fn will be public, this won't trigger clippy
provide `Program::call_meta` into public interface